### PR TITLE
[SPARK-36058][K8S] Add support for statefulset APIs in K8s

### PR DESCRIPTION
**Pull request by holdenk**
### What changes were proposed in this pull request?

Generalize the pod allocator and add support for statefulsets.

### Why are the changes needed?

Allocating individual pods in Spark can be not ideal for some clusters and using higher level operators like statefulsets and replicasets can be useful.

### Does this PR introduce _any_ user-facing change?

Yes new config options.

### How was this patch tested?

Completed: New unit & basic integration test
PV integration tests

### CODE REVIEW

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 1]
```diff
@@ -260,16 +267,22 @@ private[spark] class BasicExecutorFeatureStep(
         .withUid(pod.getMetadata.getUid)
         .build()
     }
+
+    val policy = kubernetesConf.get(KUBERNETES_ALLOCATION_PODSALLOCATOR) match {
+      case "statefulset" => "Always"
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: Well, I know that StatefulSet requires this, but does this work robustly on the loss of broadcasted variables?

File: examples/src/main/scala/org/apache/spark/examples/MiniReadWriteTest.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+import java.io.File
+
+import scala.io.Source._
+
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.util.Utils
+
+/**
+ * Simple test for reading from a consistent FS
```
Comment (dongjoon-hyun) [inline examples/src/main/scala/org/apache/spark/examples/MiniReadWriteTest.scala:29]: Do we have a test in `examples` directory?

Comment (holdenk) [reply examples/src/main/scala/org/apache/spark/examples/MiniReadWriteTest.scala:29]: Ish? It matches the location of DFS test it's based on.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: Not yet. I'm going to take a look at that after I sort out the PV integration test.

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: So I did some thinking about this (still working on the PV integration test but the code-compile-test cycle is slow), looking at `TorrentBroadcast.scala & `BroadcastManager.scala` if an executor is restarted we'll come back up with an empty `cachedValues` and then if there are any broadcast variables referenced we'll just end up fetching them anew.

We would then run into problems because we'll end up trying to store the block and the disk block manager might say it contains the block when go to do the write.

There are a few ways I can think of addressing this

1) Have the reads "fall through" (e.g. check to see if there present on disk regardless of what we know about in the block manager)
1) Ignore existing file when putting a broadcast block and just hard-write it
1) On startup (in the shell script for the dockerfile) cleanup the contents of the disk blocks location

Personally I'm tempted to do the cleanup approach, since we might have bounced for an OOM or something and I don't want to read back a partially written block. WDYT @dongjoon-hyun ?

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: I was doing more thinking about this and I think #2 (ignore existing and overwrite) is probably the most reasonable, but #1 could also be ok if we changed the behaviour of the DiskStore to do temporary file writes w/atomic renames on write completion.

Personally I'm in favour of going for #2 (ignore existing and overwrite) and leaving the temp file write w/atomic rename as a future improvement.

Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: Created https://issues.apache.org/jira/browse/SPARK-36330 to track the potential follow-up.

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: I agree with you with option 3.
> 3. On startup (in the shell script for the dockerfile) cleanup the contents of the disk blocks location
> Personally I'm tempted to do the cleanup approach

Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: BTW, `SPARK_LOCAL_DIRS` already exists? Is there a chance which the block managers pick up new storage directory when they are restarted?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 3]
```diff
@@ -323,6 +323,15 @@ private[spark] object Config extends Logging {
       .stringConf
       .createOptional
 
+  val KUBERNETES_ALLOCATION_PODSALLOCATOR =
+    ConfigBuilder("spark.kubernetes.allocation.podsallocator")
+      .doc("Allocator to use for pods. Possible values are direct (the default) and statefulset." +
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:328]: nit. We need a space at the end of this string. `statefulset."` -> `statefulset. "`.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: Ok, I'll update the startup script :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala [hunk 4]
```diff
@@ -260,16 +267,22 @@ private[spark] class BasicExecutorFeatureStep(
...
+      case _ => "Never"
+    }
     val executorPodBuilder = new PodBuilder(pod.pod)
       .editOrNewMetadata()
         .withName(name)
         .addToLabels(kubernetesConf.labels.asJava)
+        .addToLabels(SPARK_RESOURCE_PROFILE_ID_LABEL, resourceProfile.id.toString)
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: This looks like an orthogonal and good addition. Could you make a separate and simple PR for this?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/AbstractPodsAllocator.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.scheduler.cluster.k8s
+
+import io.fabric8.kubernetes.api.model.Pod
+
+import org.apache.spark.resource.ResourceProfile
+
+private[spark] abstract class AbstractPodsAllocator {
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/AbstractPodsAllocator.scala:23]: cc @shrutig since this is another `private[spark]` and will change `ExecutorPodsAllocator` [here](https://github.com/apache/spark/pull/33508/files#diff-62d442789d0ae44ae3b6ad85e82559a37dbb5b1f8a96c6dc019bc42440dc579aR46).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 6]
```diff
@@ -61,7 +61,22 @@ object ExecutorPodsSnapshot extends Logging {
 
   private def toStatesByExecutorId(executorPods: Seq[Pod]): Map[Long, ExecutorPodState] = {
     executorPods.map { pod =>
-      (pod.getMetadata.getLabels.get(SPARK_EXECUTOR_ID_LABEL).toLong, toState(pod))
+      pod.getMetadata.getLabels.get(SPARK_EXECUTOR_ID_LABEL) match {
+        case "EXECID" | null =>
+          // The pod has been created by something other than Spark and we should process
+          // pod the name to get the ID instead of the label.
+          val execIdRE = ".+-([0-9]+)".r
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:68]: Given the assumption, this RE looks fragile to me. What happens if the assumption is broken?
> The pod has been created by something other than Spark

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala [hunk 7]
```diff
@@ -61,7 +61,22 @@ object ExecutorPodsSnapshot extends Logging {
...
+          val podName = pod.getMetadata.getName
+          podName match {
+            case execIdRE(id) =>
+              (id.toLong, toState(pod))
+            case _ =>
+              throw new Exception(s"Failed to parse podname ${podName}")
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:74]: Let's use a specific exception instead of `Exception`. At least, `SparkException` is better?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 8]
```diff
@@ -323,6 +323,15 @@ private[spark] object Config extends Logging {
...
+        "Future version may add Job or replicaset.")
+      .version("3.3.0")
+      .stringConf
+      .checkValues(Set("direct", "statefulset"))
```
Comment (dongjoon-hyun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:332]: Instead of this, shall we make this more extensible by using a class name here?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: So I actually do use this in this PR, it's so that we can tell the difference between which resource profile a stateful sets pods belong to (I use it in `addToMatchLabels`).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 7 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:74]: Sure

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 6 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:68]: So if a pod is created without an exec label and the podname doesn't let us infer the exec ID we'll just throw an exception since there is not a way I could think of to recover gracefully.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: Yes, I know. What I mean is that we can merge `.addToLabels(SPARK_RESOURCE_PROFILE_ID_LABEL, resourceProfile.id.toString)` PR independency from this statefulset API.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:68]: I'm worrying about picking up a non-execID number. What happens there are two numbers fields in those strings?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: Ah ok, yeah I'll do that :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 6 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:68]: So the tests actually do that currently. Since the prefix (`.+`) is "greedy" ( https://docs.oracle.com/javase/tutorial/essential/regex/quant.html ) matching it'll only match the last numbers after the -.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 8 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:332]: I was going back and forth on that, do you think it's worthwhile to expose this as a developer API or just within Spark?

Comment (aaruna) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:332]: +1 for developer API so as to enable the usage of other options (like replicaset or Job) as applicable to other developers.

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh [hunk 9]
```diff
@@ -79,6 +79,14 @@ case "$1" in
     ;;
   executor)
     shift 1
+    # If the execid is set to the generic EXECID parse the pod name
+    if [ "${SPARK_EXECUTOR_ID}" == "EXECID" ]; then
```
Comment (aaruna) [inline resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:83]: Instead of relying on pod-name, are there other options that you have explored, like having the driver mint out executor ids upon registration?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: Dug through the code looks like this is a duplicate change already handled in KubeConf so I'll drop it.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala (hunk 8 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:332]: Ok works for me :)

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh (hunk 9 above)
Comment (holdenk) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:83]: That didn't seem to be a good match since our executor snapshot depends on being able to determine the exec id for a given pod as well and if we did it fully dynamically we wouldn't be able to know and we'd have to make even bigger changes to how we handle snapshot updates.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 1 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:272]: As discussed off-line I switched the solution here to just avoid reusing the execids so we don't need to worry about that.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala (hunk 6 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsSnapshot.scala:69]: Given our conversation around dynamically applying the label I'll do that instead and we can drop the re :)

File: resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh (hunk 9 above)
Comment (holdenk) [reply resource-managers/kubernetes/docker/src/main/dockerfiles/spark/entrypoint.sh:83]: So we had a call about this, and we'll handle it by mutating the labels on the pods.

**Review by dongjoon-hyun — COMMENTED** [review]
Could you split `GenerateExecID`-related stuff into a new PR?

**Comment by holdenk** [comment]
I could split it off into a separate PR, but if we do it as part of this PR it can be covered by the integration tests as well as the unit tests so I'd rather keep it together.

**Comment by holdenk** [comment]
K8s failure is some unrelated R stuff:

`
--- failed re-building ‘sparkr-vignettes.Rmd’

SUMMARY: processing the following file failed:
  ‘sparkr-vignettes.Rmd’
`

**Comment by holdenk** [comment]
K8s failure is some unrelated R stuff:

`
--- failed re-building ‘sparkr-vignettes.Rmd’

SUMMARY: processing the following file failed:
  ‘sparkr-vignettes.Rmd’
`

**Comment by holdenk** [comment]
Jenkins retest this please.

**Comment by holdenk** [comment]
If there are no other concerns, I plan to merge this early next week (current dev branch only).

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala [hunk 10]
```diff
@@ -323,6 +323,16 @@ private[spark] object Config extends Logging {
       .stringConf
       .createOptional
 
+  val KUBERNETES_ALLOCATION_PODSALLOCATOR =
+    ConfigBuilder("spark.kubernetes.allocation.podsallocator")
```
Comment (kbendick) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:327]: Nit: Possibly this config would be better named as `spark.kubernetes.allocation.pods.allocator`? Seems that `Allocator` is used as its own word and is capitalized in class names.

I'd also update to `KUBERNETES_ALLOCATION_PODS_ALLOCATOR` if you make that change.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala [hunk 11]
```diff
@@ -138,6 +133,31 @@ private[spark] class KubernetesClusterManager extends ExternalClusterManager wit
       podsPollingEventSource)
   }
 
+  private[k8s] def makeExecutorPodsAllocator(sc: SparkContext, kubernetesClient: KubernetesClient,
+      snapshotsStore: ExecutorPodsSnapshotsStore) = {
+    val executorPodsAllocatorName = sc.conf.get(KUBERNETES_ALLOCATION_PODSALLOCATOR) match {
+      case "statefulset" =>
+        "org.apache.spark.scheduler.cluster.k8s.StatefulsetPodsAllocator"
+      case "direct" =>
+        "org.apache.spark.scheduler.cluster.k8s.ExecutorPodsAllocator"
```
Comment (kbendick) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala:142]: Nit: Possibly consider using some named constants or `classOf` with `.toString`? Can be left as a follow up too.

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocatorSuite.scala [hunk 12]
```diff
@@ -242,6 +246,17 @@ class ExecutorPodsAllocatorSuite extends SparkFunSuite with BeforeAndAfter {
     verify(podOperations).create(podWithAttachedContainerForId(podAllocationSize + 1))
   }
 
+  test("Verify stopping deletes the labled pods") {
```
Comment (kbendick) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocatorSuite.scala:249]: Nit: typo in the word `labeled`.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/BasicTestsSuite.scala [hunk 13]
```diff
@@ -25,11 +29,24 @@ private[spark] trait BasicTestsSuite { k8sSuite: KubernetesSuite =>
...
   import BasicTestsSuite._
   import KubernetesSuite.k8sTestTag
+  import KubernetesSuite.{TIMEOUT, INTERVAL}
 
   test("Run SparkPi with no resources", k8sTestTag) {
     runSparkPiAndVerifyCompletion()
   }
 
+  test("Run SparkPi with no resources & statefulset allocation", k8sTestTag) {
+    sparkAppConf.set("spark.kubernetes.allocation.podsallocator", "statefulset")
```
Comment (kbendick) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/BasicTestsSuite.scala:39]: If you update the config name, don't forget to update here too =)

**Review by kbendick — COMMENTED** [review]
A few nits I'll leave at your discretion, but overall this looks good to me for supporting statefulsets and other executor allocation strategies. Would love to get this in to make testing the use of this API easier. 🙂

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala (hunk 11 above)
Comment (holdenk) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/KubernetesClusterManager.scala:142]: Oh hells yeah, good call.

**Comment by holdenk** [comment]
jenkins retest this please.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala (hunk 4 above)
Comment (Yikun) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/BasicExecutorFeatureStep.scala:279]: Remove it in : https://github.com/apache/spark/pull/34646

**Comment by yipen** [comment]
@holdenk I noticed that there already set ownerreference between the executor pods and driver pod. And there also set ownerreference between the statefulset and dirver. Not sure if these are duplicated?
